### PR TITLE
Load the Cart Fragments script on specific pages

### PR DESCRIPTION
**Pull request by tarhi-saad**
In an effort to improve the JS performance for the Cart Fragments (see P2: pdFofs-162-p2), we want to load the `cart-fragments` JS script selectively:

- **What we already do:** The `cart-fragments` is only loaded when the Legacy Cart Widget is used. But it's loaded on all pages of the site. Even on pages that would not update the cart (e.g., Home). See [line](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/widgets/class-wc-widget-cart.php#L59).
- **What the PR is changing:** We want to load the `cart-fragments` script only on pages that would update the cart (see P2's suggestion: pdFofs-162-p2#comment-1330):
  - So, in this [line](https://github.com/woocommerce/woocommerce/blob/try/load-cart-fragments-script-on-specific-pages/plugins/woocommerce/includes/widgets/class-wc-widget-cart.php#L112-L113), we are loading the script only for specific pages and blocks that allow updating the cart.
  - Since the cart-fragments script isn’t loaded on other pages (e.g., home), an empty Cart Widget is displayed (i.e., we don’t get the cart data). To fix this bug, we send the cart content from the back-end on page load (see [line](https://github.com/woocommerce/woocommerce/blob/try/load-cart-fragments-script-on-specific-pages/plugins/woocommerce/includes/widgets/class-wc-widget-cart.php#L116-L119))

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes # .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Add the Legacy Cart Widget to the sidebar
2. Go to the shop page.
3. Add one or more products to the cart. Ensure the Cart Widget is updated accordingly
4. Go to the home page. Ensure the Cart Widget is correctly displayed

### CODE REVIEW

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php [hunk 1]
```diff
@@ -70,8 +106,17 @@ public function widget( $args, $instance ) {
...
+		$is_cart_related_block = $this->array_contains( $post->post_content, self::get_blocks() );
+
+		if ( $is_cart_related_block || is_shop() || is_product() || is_product_category() ) {
+			wp_enqueue_script( 'wc-cart-fragments' );
+			// Insert cart widget placeholder - code in woocommerce.js will update this on page load.
+			echo '<div class="widget_shopping_cart_content"></div>';
+		} else {
+			// Insert cart widget with the Mini-cart content.
+			echo '<div class="widget_shopping_cart_content">' . woocommerce_mini_cart() . '</div>';
+		}
```
Comment (senadir) [inline plugins/woocommerce/includes/widgets/class-wc-widget-cart.php]: If `wc-cart-fragments` will fetch `woocommerce_mini_cart` over the wire anyway, why not just always print it, and only enqueue the script if needed?

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php [hunk 2]
```diff
@@ -43,6 +43,44 @@ public function __construct() {
...
+	}
+
+	/**
+	 * array_contains
+	 * Check if string contains word from array.
+	 *
+	 * @since 1.0.0
+	 * @version 1.0.0
+	 **/
+	function array_contains( $str, array $arr ) {
```
Comment (senadir) [inline plugins/woocommerce/includes/widgets/class-wc-widget-cart.php]: It would be nice to reshape this to resemble `has_block`, so something like `has_blocks( srtring[] $blocks, int|string|WP_Post|null $post)`

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php [hunk 3]
```diff
@@ -43,6 +43,44 @@ public function __construct() {
...
+			'woocommerce/all-products',
+		);
+	}
+
+	/**
+	 * array_contains
+	 * Check if string contains word from array.
+	 *
+	 * @since 1.0.0
+	 * @version 1.0.0
```
Comment (senadir) [inline plugins/woocommerce/includes/widgets/class-wc-widget-cart.php]: Linting for function fails because it has no PHPDocs for the params.

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php (hunk 2 above)
Comment (senadir) [inline plugins/woocommerce/includes/widgets/class-wc-widget-cart.php]: Function needs a private visibility param.

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php [hunk 4]
```diff
@@ -70,8 +106,17 @@ public function widget( $args, $instance ) {
...
+		global $post;
+		$is_cart_related_block = $this->array_contains( $post->post_content, self::get_blocks() );
+
+		if ( $is_cart_related_block || is_shop() || is_product() || is_product_category() ) {
+			wp_enqueue_script( 'wc-cart-fragments' );
+			// Insert cart widget placeholder - code in woocommerce.js will update this on page load.
+			echo '<div class="widget_shopping_cart_content"></div>';
+		} else {
+			// Insert cart widget with the Mini-cart content.
+			echo '<div class="widget_shopping_cart_content">' . woocommerce_mini_cart() . '</div>';
```
Comment (senadir) [inline plugins/woocommerce/includes/widgets/class-wc-widget-cart.php]: woocommerce_mini_cart should be escaped with `esc_html`

**Review by senadir — CHANGES_REQUESTED** [review]
I left some questions and refactoring request to pass tests.

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php (hunk 1 above)
Comment (tarhi-saad) [reply plugins/woocommerce/includes/widgets/class-wc-widget-cart.php]: @senadir, the `wc-cart-fragments` will fetch `woocommerce_mini_cart` for the first time only. After that, it will get the cart data from the Browser's session storage if the cart content didn't get updated between page loads (see [line](https://github.com/woocommerce/woocommerce/blob/try/load-cart-fragments-script-on-specific-pages/plugins/woocommerce/client/legacy/js/frontend/cart-fragments.js#L144-L148)). I believe there is a performance benefit to not print the cart content on each page load for pages that loads the `wc-cart-fragments` script. Please let me know if you think otherwise! 🙌

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php (hunk 2 above)
Comment (tarhi-saad) [reply plugins/woocommerce/includes/widgets/class-wc-widget-cart.php]: I added a fix in this [commit](https://github.com/woocommerce/woocommerce/pull/38910/commits/dc47b563c013755ae85e797419449aa627002e84). I named the function `has_block_from_list` since `has_blocks` already exist!

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php (hunk 3 above)
Comment (tarhi-saad) [reply plugins/woocommerce/includes/widgets/class-wc-widget-cart.php]: Nice catch! I fixed it in the new function! 🙌

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php (hunk 2 above)
Comment (tarhi-saad) [reply plugins/woocommerce/includes/widgets/class-wc-widget-cart.php]: Thank you! It's added!

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php (hunk 4 above)
Comment (tarhi-saad) [reply plugins/woocommerce/includes/widgets/class-wc-widget-cart.php]: Done!

**Comment by tarhi-saad** [comment]
Thank you, @senadir, for the review! I addressed the above suggestion! Could you give this PR a second look?

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php [hunk 5]
```diff
@@ -43,6 +43,60 @@ public function __construct() {
...
+
+		$block_from_page = parse_blocks( $post->post_content );
+		$block_name      = $block_from_page[0]['blockName'];
+
+		if ( in_array( $block_name, $blocks, true ) ) {
+			return true;
+		}
+
+		return false;
+	}
```
Comment (senadir) [inline plugins/woocommerce/includes/widgets/class-wc-widget-cart.php]: This logic doesn't make sense to me, why are we only checking the first block in the page? what if the first block was a generic core block?

Comment (senadir) [reply plugins/woocommerce/includes/widgets/class-wc-widget-cart.php]: I think the previous logic of using strpos and mapping over the array (and breaking early) was good enough? it's the same logic that has_block uses.

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php (hunk 1 above)
Comment (ralucaStan) [reply plugins/woocommerce/includes/widgets/class-wc-widget-cart.php]: @Aljullu might have more context on the Mini cart loading.

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php [hunk 6]
```diff
@@ -43,6 +43,60 @@ public function __construct() {
 		parent::__construct();
 	}
 
+	/**
+	 * Get list of Woo blocks that update the cart.
+	 *
+	 * @return array;
+	 */
+	public static function get_blocks() {
```
Comment (ralucaStan) [inline plugins/woocommerce/includes/widgets/class-wc-widget-cart.php]: `get_blocks_that_update_the_cart` as a name function would make the comment obsolete.

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php [hunk 7]
```diff
@@ -43,6 +43,60 @@ public function __construct() {
...
+	/**
+	 * Check if post has a block from the provided list of blocks.
+	 *
+	 * @param string[]                $blocks     Blocks to check for.
+	 * @param int|string|WP_Post|null $post       Post to check.
+	 *
+	 * @since 1.0.0
+	 * @version 1.0.0
+	 **/
+	private function has_block_from_list( $blocks, $post = null ) {
```
Comment (ralucaStan) [inline plugins/woocommerce/includes/widgets/class-wc-widget-cart.php:78]: I think the naming `has_block_from_list` can still be improved to reflect the intention. `has_blocks_that_update_the_cart` maybe?

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php [hunk 8]
```diff
@@ -43,6 +43,60 @@ public function __construct() {
 		parent::__construct();
 	}
 
+	/**
+	 * Get list of Woo blocks that update the cart.
+	 *
```
Comment (ralucaStan) [inline plugins/woocommerce/includes/widgets/class-wc-widget-cart.php:49]: I'd explain the intention behind this curated list to leave out any confusion. 
One might ask himself why aren't the Cart and Checkout blocks included, as they can update the cart.

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php (hunk 7 above)
Comment (tarhi-saad) [reply plugins/woocommerce/includes/widgets/class-wc-widget-cart.php:78]: If the function name is `has_blocks_that_update_the_cart`, then it should accept `$post` as the only argument. It won't make sense to allow providing any other list of blocks. As a result, we should have something like this:

```PHP
private function has_blocks_that_update_the_cart( $post = null ) {
```

But, I find the following line explains the function implementation better than the last one:

```PHP
$is_cart_related_block = $this->has_block_from_list( self:: get_blocks_that_update_the_cart(), $post );
```

```PHP
$is_cart_related_block = $this->has_blocks_that_update_the_cart( $post );
```
With `has_blocks_that_update_the_cart`, we won't understand how we get these blocks until we check the function's body.

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php (hunk 5 above)
Comment (tarhi-saad) [reply plugins/woocommerce/includes/widgets/class-wc-widget-cart.php]: Nice catch, @senadir! You are right, `has_block` use a similar logic. I thought I was using a hacky solution!

**Comment by tarhi-saad** [comment]
Thank you for your review, @senadir, @ralucaStan! I updated the PR. Please let me know if I missed anything! 🙌

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php (hunk 7 above)
Comment (ralucaStan) [reply plugins/woocommerce/includes/widgets/class-wc-widget-cart.php:78]: I see, this is a general function that tests for any blocks. Thank you for explaining. Makes sense.

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php [hunk 9]
```diff
@@ -43,6 +43,56 @@ public function __construct() {
...
+			'woocommerce/add-to-cart-form',
+			'woocommerce/product-top-rated',
+			'woocommerce/product-tag',
+			'woocommerce/products-by-attribute',
+			'woocommerce/product-on-sale',
+			'woocommerce/product-new',
+			'woocommerce/product-category',
+			'woocommerce/product-best-sellers',
+			'woocommerce/all-products',
+		);
```
Comment (Aljullu) [inline plugins/woocommerce/includes/widgets/class-wc-widget-cart.php:66]: Should we make this list filterable? Ie: what if a plugin wants to create a new block that allows updating the Cart?

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php [hunk 10]
```diff
@@ -70,8 +118,17 @@ public function widget( $args, $instance ) {
 			echo '<div class="hide_cart_widget_if_empty">';
 		}
 
-		// Insert cart widget placeholder - code in woocommerce.js will update this on page load.
-		echo '<div class="widget_shopping_cart_content"></div>';
+		global $post;
+		$is_cart_related_block = $this->has_block_from_list( self::get_blocks_that_update_the_cart(), $post );
+
+		if ( $is_cart_related_block || is_shop() || is_product() || is_product_category() ) {
```
Comment (Aljullu) [inline plugins/woocommerce/includes/widgets/class-wc-widget-cart.php:124]: Should we also check for shortcodes, here?

File: plugins/woocommerce/includes/widgets/class-wc-widget-cart.php [hunk 11]
```diff
@@ -70,8 +118,17 @@ public function widget( $args, $instance ) {
...
+			wp_enqueue_script( 'wc-cart-fragments' );
+			// Insert cart widget placeholder - code in woocommerce.js will update this on page load.
+			echo '<div class="widget_shopping_cart_content"></div>';
+		} else {
+			// Insert cart widget with the Mini-cart content.
+			echo '<div class="widget_shopping_cart_content">' . esc_html( woocommerce_mini_cart() ) . '</div>';
```
Comment (Aljullu) [inline plugins/woocommerce/includes/widgets/class-wc-widget-cart.php:130]: Heads-up that if we print the Mini-Cart contents directly from the PHP code, that will cause wrong data to be displayed if the store has installed caching plugins like [WP-Optimize](https://wordpress.org/plugins/wp-optimize/).

**Review by Aljullu — COMMENTED** [review]
While I understand the intention to try to not load Cart Fragments when possible, I'm not sure whether that's something achievable with the way WooCommerce works right now. The main issues I think we have:

1. It's quite hard to predict whether the page allows modifying the cart or not. We can check which page it is, and whether it has certain blocks or not (as this PR does), but we should also check for certain shortcodes. And I guess plugins or themes might be introducing other ways to modify the cart from certain pages that we might not be aware of, so I think this is quite risky.
2. The second issue is that if we want to print Mini-Cart data from the PHP-side, that won't work well with caching plugins. We recently changed how the Mini-Cart block works to fix that (https://github.com/woocommerce/woocommerce-blocks/pull/9493).

To summarize, I'm not sure we can avoid loading Cart Fragments without introducing other issues which IMO are blockers.

---

Having said that, I'm fine with whatever decision we take. :slightly_smiling_face: I took a look at this PR and overall it looks well, but found some issues (mostly related to what I commented above), but mentioning them more concretely below: :point_down: 

If I try to remove a product from the cart from the My Account page, it seems to append the new cart markup after the existing one. Can you reproduce as well? Tested on Storefront with the Mini Cart widget on the sidebar.

[Enregistrament de pantalla des de 2023-07-10 13-47-56.webm](https://github.com/woocommerce/woocommerce/assets/3616980/21e68a7b-f850-4c09-8300-b758365c1545)

**Comment by ralucaStan** [comment]
> While I understand the intention to try to not load Cart Fragments when possible, I'm not sure whether that's something achievable with the way WooCommerce works right now. The main issues I think we have:
> 
> 1. It's quite hard to predict whether the page allows modifying the cart or not. We can check which page it is, and whether it has certain blocks or not (as this PR does), but we should also check for certain shortcodes. And I guess plugins or themes might be introducing other ways to modify the cart from certain pages that we might not be aware of, so I think this is quite risky.
> 2. The second issue is that if we want to print Mini-Cart data from the PHP-side, that won't work well with caching plugins. [...]
> 
> To summarize, I'm not sure we can avoid loading Cart Fragments without introducing other issues which IMO are blockers.
> 

Thank you @Aljullu  for pointing out these important aspects. My main takeaways are:
- it's hard to implement a bulletproof solution to detect what can update the cart on a page.
- this PR is looking at WooCommerce-owned blocks that update the cart, but there can be shortcodes or themes with this logic
- printing Mini Cart data from PHP side for the first load, which is a must for removing the script on some pages, is error prone for caching reasons. 

My thoughts are that maybe this approach is not the best one to take because, outside of the reasons specified by you, there are also maintenance reasons. For example, adding a new Woo block that updates the cart should be recorded here as well, and I feel this is something that can be easily overlooked. 

We could change the approach in the sense that we collect all the official ways a cart can be updated using WooCommerce blocks & shortcodes or even filters, as seen [here](https://github.com/woocommerce/woocommerce/pull/38736). Aka things we own, control, and can determine. And disabling the cart fragments for pages that contain these WooCommerce-owned elements can be an opt-in feature, as in people choose to activate it depending on their setup. But this also does not fix the issues with the Mini Cart caching. 

But I wonder if after @mikejolley 's PR https://github.com/woocommerce/woocommerce/pull/35530 we are still dealing with performance issue. I personally feel it's not worth bringing in this extra level of complexity.

**Comment by tarhi-saad** [comment]
Thank you, @Aljullu, @ralucaStan, for your feedback! I've raised similar concerns in the related P2 (pdFofs-162-p2#comment-1398):

> The current approach has certain drawbacks. For instance, if we introduce a new block that can update the cart, we must update the block list in the Cart Fragments script loading logic. Additionally, some third-party extensions may provide the option to edit the cart on specific pages. In cases where these pages didn’t initially have this feature, the Cart Widget will not receive live updates.

We also can't measure the performance improvement for this change with certainty. The server response is unlikely to be substantially affected, except perhaps during periods of high traffic where it could potentially be impacted.

> I personally feel it's not worth bringing in this extra level of complexity.

I've shared the same thought in a previous exchange. I anticipated that the work ahead might be more complex and substantial compared to the benefits we will gain.

The decision will depend on the following: 
 We can continue pursuing this if you consider the expected performance boost to sufficiently justify the associated work and uncertainties. However, if you feel the additional complexity is not justified by the potential advantages, we can close this Pull Request!

**Comment by ralucaStan** [comment]
I think it's fine to close this PR.
